### DISCUSSION TITLE
How to generate a pore network model with different pore distribution in the upper and lower layers?

### DISCUSSION THREAD

**Discussion by Xknightfyy** [PMEAL/OpenPNM#2750] [Q&A]
Dear Professor Jeff Gostick,

I hope this email finds you well. Thank you for your previous responses. I am reaching out to you regarding a new issue I have encountered.

I am trying to generate two models with different pore distributions in a network model using the stitch function. I followed your tutorial on Calibrating a Network Using Mercury Intrusion Data on the website for pore distribution. However, even after using pn.regenerate_models(), I am still unable to obtain the model with different pore distributions in the upper and lower layers.

Could you please advise me on how to address this issue? Your prompt response would be greatly appreciated.

Best regards,

xu

**Comment by Xknightfyy** [comment]
import openpnm as op
import numpy as np
import math as m
import os
current_directory = os.getcwd()
import matplotlib.pyplot as plt
import scipy.stats as spst
op.visualization.set_mpl_style()
np.random.seed(10000)

pn = op.network.Cubic(shape=[20,20,5], spacing=[1e-6,1e-6,1e-6])
pn1 = op.network.Cubic(shape=[20,20,5], spacing=[1e-6,1e-6,1e-6])
pn1.add_model_collection(op.models.collections.geometry.spheres_and_cylinders)
pn1.regenerate_models()
pn.add_model_collection(op.models.collections.geometry.spheres_and_cylinders)
pn.regenerate_models()

np.random.seed(0)  # Set the state of the random number generator to "0"
pn['pore.seed'] = np.random.rand(pn.Np)
dst = spst.weibull_min(c=1.9, loc=1.78e-5, scale=5e-8)
pn['pore.diameter'] = dst.rvs(pn.Np)
f = op.models.geometry.pore_seed.random
pn.add_model(propname='pore.seed',
             model=f,
             num_range=[0.01, 0.99])
pn.add_model(propname='pore.diameter',
             model=op.models.geometry.pore_size.generic_distribution,
             func=dst,)
pn.add_model(propname='throat.diameter_1', 
             model=op.models.misc.from_neighbor_pores,
             prop='pore.diameter',
             mode='min')
pn.add_model(propname='throat.seed', 
             model=op.models.misc.from_neighbor_pores,
             prop='pore.seed',
             mode='min')
pn.add_model(propname='throat.diameter_2',
             model=op.models.geometry.throat_size.generic_distribution,
             func=dst)
pn.add_model(propname='throat.diameter',
             model=op.models.misc.scaled,
             prop='throat.diameter_1',  # This could also be 'throat.diameter_2'
             factor=0.6,  # This could be 1.0 if no scaling is desired
             )
pn.regenerate_models()
pn.add_model(propname='pore.volume',
             model=op.models.geometry.pore_volume.sphere)
pn.add_model(propname='throat.length',
             model=op.models.geometry.throat_length.spheres_and_cylinders)
pn.add_model(propname='throat.total_volume',
             model=op.models.geometry.throat_volume.cylinder)
pn.add_model(propname='throat.lens_volume', 
            model=op.models.geometry.throat_volume.lens)
pn.add_model(propname='throat.volume', 
             model=op.models.misc.difference,
             props=['throat.total_volume', 'throat.lens_volume'])

fig, ax = plt.subplots()
ax.hist(pn['pore.diameter']*1e6, edgecolor='k')
ax.tick_params(axis='x', labelrotation=45)
ax.set_xlabel('Pore Diameter [um]');

np.random.seed(0)  # Set the state of the random number generator to "0"
pn1['pore.seed'] = np.random.rand(pn1.Np)
dst = spst.weibull_min(c=1.9, loc=1.78e-6, scale=5e-8)
pn1['pore.diameter'] = dst.rvs(pn.Np)
f = op.models.geometry.pore_seed.random
pn1.add_model(propname='pore.seed',
             model=f,
             num_range=[0.01, 0.99])
pn1.add_model(propname='pore.diameter',
             model=op.models.geometry.pore_size.generic_distribution,
             func=dst,)
pn1.add_model(propname='throat.diameter_1', 
             model=op.models.misc.from_neighbor_pores,
             prop='pore.diameter',
             mode='min')
pn1.add_model(propname='throat.seed', 
             model=op.models.misc.from_neighbor_pores,
             prop='pore.seed',
             mode='min')
pn1.add_model(propname='throat.diameter_2',
             model=op.models.geometry.throat_size.generic_distribution,
             func=dst)
pn1.add_model(propname='throat.diameter',
             model=op.models.misc.scaled,
             prop='throat.diameter_1',  # This could also be 'throat.diameter_2'
             factor=0.6,  # This could be 1.0 if no scaling is desired
             )
pn1.regenerate_models()
pn1.add_model(propname='pore.volume',
             model=op.models.geometry.pore_volume.sphere)
pn1.add_model(propname='throat.length',
             model=op.models.geometry.throat_length.spheres_and_cylinders)
pn1.add_model(propname='throat.total_volume',
             model=op.models.geometry.throat_volume.cylinder)
pn1.add_model(propname='throat.lens_volume', 
            model=op.models.geometry.throat_volume.lens)
pn1.add_model(propname='throat.volume', 
             model=op.models.misc.difference,
             props=['throat.total_volume', 'throat.lens_volume'])

fig, ax = plt.subplots()
ax.hist(pn1['pore.diameter']*1e6, edgecolor='k')
ax.tick_params(axis='x', labelrotation=45)
ax.set_xlabel('Pore Diameter [um]');

z_move= 6e-6
pn['pore.coords'] += [ 0,  0, z_move]
op.topotools.stitch(network=pn, donor=pn1, P_network=pn.pores('bottom'), P_donor=pn1.pores('top'))

ax = op.visualization.plot_connections(pn, color_by=pn['throat.diameter'], linewidth=4)
ax = op.visualization.plot_coordinates(pn, size_by=pn['pore.diameter'], markersize=30, ax=ax)


Here is my code part, can you help me see where the problem is?

**Reply by jgostick** [reply]
In V3 we 'tried' to simplify OpenPNM, but front and backend, by removing the Geometry and Physics objects, and instead you put all the models/data directly on network and phase objects.  The trade off with this new approach is that when you add a model, you need to tell it 'where' it's results should go.  We do this using labels.  When I look at your code it seems that you are overwriting the first models when you put the second ones in.  Please have a look at [this tutorial/blog post](http://pmeal.com/posts/2022-10-10-notebook-post/) for some hints.

**Reply by Xknightfyy** [reply]
If I do this, will the size of the pore network model change after pn.regenerate_models ()? In the next operation, I must use pn.regenerate_models ().

**Reply by jgostick** [reply]
no, the 'domain' argument in `add_model` just tells OpenPNM "where" the results from the model should go.  The default is "all", but you can do "left" or any other custom label, like "layer1" and "layer2".

**Reply by Xknightfyy** [reply]
Wow! Thank you for your reply. It helped me solve the problem and I really appreciate it.
